src/pe/relocation.rs

```diff
@@ -131,3 +308,364 @@ impl<'a> Iterator for Relocations<'a> {
         }
     }
 }
+
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct RelocationData<'a> {
+    bytes: &'a [u8],
+}
+
+impl<'a> RelocationData<'a> {
+    pub fn parse(
+        bytes: &'a [u8],
+        dd: data_directories::DataDirectory,
+        sections: &[section_table::SectionTable],
+        file_alignment: u32,
+    ) -> error::Result<Self> {
+        Self::parse_with_opts(
+            bytes,
+            dd,
+            sections,
+            file_alignment,
+            &options::ParseOptions::default(),
+        )
+    }
+
+    pub fn parse_with_opts(
+        bytes: &'a [u8],
+        dd: data_directories::DataDirectory,
+        sections: &[section_table::SectionTable],
+        file_alignment: u32,
+        opts: &options::ParseOptions,
+    ) -> error::Result<Self> {
+        let offset =
+            utils::find_offset(dd.virtual_address as usize, sections, file_alignment, opts)
+                .ok_or_else(|| {
+                    error::Error::Malformed(format!(
+                        "Cannot map base reloc rva {:#x} into offset",
+                        dd.virtual_address
+                    ))
+                })?;
+
+        // Ensure that the offset and size do not exceed the length of the bytes slice
+        if offset + dd.size as usize > bytes.len() {
+            return Err(error::Error::Malformed(format!(
+                "base reloc offset {:#x} and size {:#x} exceeds the bounds of the bytes size {:#x}",
+                offset,
+                dd.size,
+                bytes.len()
+            )));
+        }
+
+        let bytes = &bytes[offset..offset + dd.size as usize];
```

Comment:
it doesn't matter now but i wonder if a better pattern here to use is something like:

```rust
let size = dd.size as usize;
let bytes = bytes[offset..].pread::<&[u8]>(size).map_err(|_| TheErrorWeWantToDisplay)?;
```

it's just easy to copy-pasta the bounds error, and slice the wrong bounds that were different than what we checked, etc. anyway, not too important.

Reply:
Wow, didn't know that there's such a great expression. Thanks!